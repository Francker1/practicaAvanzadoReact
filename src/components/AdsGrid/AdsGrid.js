import React from "react";
import T from "prop-types";
import { Link } from "react-router-dom";

import Username from "../_username";
import { Container, Card } from "react-bootstrap";

import ButtonFavs from "../_common/buttons/ButtonAdFavs";
import AdsList from "../AdsList";

function AdCard({
    _id,
    photo,
    name,
    price,
    description,
    type,
    tags,
    createdAt,
    updatedAt,
    onAddToFavsClick,
}) {
    return (
        <>
        <Link to={`/detail/${_id}`}>
            <Card>
                <Card.Img className="img-card" variant="top" src={photo} />
                <Card.Body>
                    <Card.Title>{name}</Card.Title>
                    <Card.Text as={"div"}>
                        <dl>
                            <dt>Precio: {price} €</dt>

                            <dt>Desc:</dt>
                            <dd>{description}</dd>

                            <dt>Tipo:</dt>
                            <dd>{type}</dd>

                            <dt>Tags:</dt>
                            {tags && tags.map(tag => <dd key={tag}>{tag}</dd>)}

                            <dt>created:</dt>
                            <dd>created: {createdAt}</dd>
                        </dl>
                    </Card.Text>
                    <Card.Footer>
                        <small className="text-muted">
                            Última actualización: {updatedAt}
                        </small>
                    </Card.Footer>
                </Card.Body>
            </Card>
            </Link>
            <ButtonFavs
                action={onAddToFavsClick} 
            />{' '}
        </>
    );
}

AdCard.propTypes = {
    _id: T.string.isRequired,
    photo: T.string,
    name: T.string.isRequired,
    price: T.number.isRequired,
    description: T.string.isRequired,
    type: T.string.isRequired,
    tags: T.array,
    createdAt: T.string.isRequired,
    updatedAt: T.string.isRequired,
};

export default function AdsGrid({ ads, addToFavs }) {
    return (
        <Container>
            <Username/>
            <AdsList
                items={ads}
                renderItem={ad => (
                    <AdCard
                        {...ad}
                        onAddToFavsClick={() => addToFavs(ad._id)}
                    />
                )}
            />
        </Container>
    );
}

AdsGrid.propTypes = {
    ads: T.arrayOf(T.shape({ _id: T.string.isRequired })),
    addToFavs: T.func.isRequired,
};
