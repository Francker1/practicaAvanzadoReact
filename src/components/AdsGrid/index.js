import { connect } from "react-redux";

import AdsGrid from "./AdsGrid";

import { addToFavs } from "../../store/actions";


function mapStateToProps(state, ownProps) {
    return {
        ads: state.ads,
    };
}

function mapDispatchToProps(dispatch, ownProps) {
    return {
        addToFavs: adId => dispatch(addToFavs(adId)),
    };
}


export default connect(mapStateToProps, mapDispatchToProps)(AdsGrid);
