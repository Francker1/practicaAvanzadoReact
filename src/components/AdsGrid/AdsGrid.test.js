import React from "react";
import { shallow } from "enzyme";
import AdsGrid from "./AdsGrid";

describe("AdsGrid", () => {
  const props = {
    ads: [
      {
        tags: ["lifestyle"],
        _id: "5e43096e99660405061db854",
        name: "incredibly awesome",
        price: 15,
        description: "test description",
        type: "sell",
        photo: "https://pbs.twimg.com/media/Dpu128FWwAI_Bru?format=jpg",
        createdAt: "2020-02-11T20:07:10.232Z",
        updatedAt: "2020-05-31T09:38:39.133Z",
      },
    ],
    addToFavs: jest.fn(),
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<AdsGrid {...props} />);
  });
  wrapper = shallow(<AdsGrid {...props} />);
  console.log(wrapper);

  test("should render ", () => {
    expect(wrapper.exists()).toBe(true);
  });

  test("snapshot testing", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should render a List of ads", () => {
    expect(wrapper.find("AdsList").props().items).toEqual(props.ads);
  });

  test("should add a advertisement to favorites", () => {
    const ad = {
      tags: ["lifestyle"],
      _id: "5e43096e99660405061db854",
      name: "incredibly awesome",
      price: 15,
      description: "test description",
      type: "sell",
      photo: "https://pbs.twimg.com/media/Dpu128FWwAI_Bru?format=jpg",
      createdAt: "2020-02-11T20:07:10.232Z",
      updatedAt: "2020-05-31T09:38:39.133Z",
    };

    const adItem = wrapper.find("AdsList").props().renderItem(ad);
    adItem.props.onAddToFavsClick();
    expect(props.addToFavs).toHaveBeenCalledWith(ad._id);
  });
});
