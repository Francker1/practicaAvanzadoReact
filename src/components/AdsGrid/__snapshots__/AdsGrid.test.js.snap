// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdsGrid snapshot testing 1`] = `
<Container
  fluid={false}
>
  <Connect(Username) />
  <AdsList
    items={
      Array [
        Object {
          "_id": "5e43096e99660405061db854",
          "createdAt": "2020-02-11T20:07:10.232Z",
          "description": "test description",
          "name": "incredibly awesome",
          "photo": "https://pbs.twimg.com/media/Dpu128FWwAI_Bru?format=jpg",
          "price": 15,
          "tags": Array [
            "lifestyle",
          ],
          "type": "sell",
          "updatedAt": "2020-05-31T09:38:39.133Z",
        },
      ]
    }
    renderItem={[Function]}
  />
</Container>
`;
